rskj-core/src/main/java/co/rsk/net/BlockSyncService.java

```diff
@@ -0,0 +1,332 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2017 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.net;
+
+import co.rsk.core.bc.BlockChainStatus;
+import co.rsk.core.bc.BlockUtils;
+import co.rsk.net.messages.*;
+import org.ethereum.core.*;
+import org.ethereum.db.ByteArrayWrapper;
+import org.ethereum.net.server.ChannelManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.spongycastle.util.encoders.Hex;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import java.math.BigInteger;
+import java.util.*;
+
+/**
+ * BlockSyncService processes blocks to add into a blockchain.
+ * If a block is not ready to be added to the blockchain, it will be on hold in a BlockStore.
+ * <p>
+ * This class is tightly coupled with
+ */
+public class BlockSyncService {
+    private static final int NBLOCKS_TO_SYNC = 30;
+
+    @GuardedBy("syncLock") private volatile int nsyncs = 0;
+    @GuardedBy("syncLock") private volatile boolean syncing = false;
```

Comment:
release of locks induce a happens-before status for the acquisition, so it sounds weird a @GuarderBy + volatile

Reply:
there's no object such as syncLock, did you mean this? as it is updated on synchronized methods

Reply:
We have been discussing about removing this lock it seems some guards were remaining.
The methods where nsyncs and syncing are used should be reimplemented later.